test/integration/integration_test.cc

```diff
@@ -2409,6 +2409,103 @@ TEST_P(IntegrationTest, RandomPreconnect) {
   }
 }
 
+/*
+This test ensure that the feature flag: envoy.restart_features.ensure_connection_retry
+enables retries on connection timeouts even if the buffer limit has been reached.
+Also check that the router pause the reading on the downstream connection when the buffer limit was
+reached.
+*/
+TEST_P(IntegrationTest, EnsureConnectionRetry) {
+  // Set buffer limits per requests
+  config_helper_.addConfigModifier(
+      [](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+             hcm) {
+        auto* route_config = hcm.mutable_route_config();
+        auto* virtual_host = route_config->mutable_virtual_hosts(0);
+        auto* route = virtual_host->mutable_routes(0)->mutable_per_request_buffer_limit_bytes();
+        route->set_value(64);
+      });
+  // increase the base backoff to ensure we have time to see the connection failure and start
+  // the new upstream before the retry.
+  config_helper_.addRuntimeOverride("upstream.base_retry_backoff_ms", "200");
+  // enable feature flags envoy.restart_features.ensure_connection_retry
+  config_helper_.addRuntimeOverride("envoy.restart_features.ensure_connection_retry", "true");
+
+  // start Envoy and close the socket of the upstream
+  initialize();
+  Network::Address::InstanceConstSharedPtr upstream_address = fake_upstreams_[0]->localAddress();
+  fake_upstreams_[0]->closeSocket();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto encoder_decoder = codec_client_->startRequest(
+      Http::TestRequestHeaderMapImpl{{":method", "POST"},
+                                     {":path", "/test/long/url"},
+                                     {":scheme", "http"},
+                                     {":authority", "sni.lyft.com"},
+                                     {"x-forwarded-for", "10.0.0.1"},
+                                     {"x-envoy-retry-on", "connect-failure"},
+                                     {"x-envoy-max-retries", "5"}});
+
+  auto& encoder = encoder_decoder.first;
+  auto& response = encoder_decoder.second;
+  // Send more data than the buffer limit, and not end-stream
+  codec_client_->sendData(encoder, 128, false);
+
+  // waiting to have at least one connection failure
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_connect_fail", 0);
```

Comment:
shouldn't this wait for this to be 1?

Reply:
I guess this could be a no-op in some case but it guarantee us that we got at least one connection failure before creating a healthy upstream. Which is what we want to test here.
IMO it does not make the test more flaky as we allows up to 5 retries, quite the opposite as I explained above

I ran the test around 10 times on my local machine without any issue

Reply:
it guarantee us that we got at least one connection failure before creating a healthy upstream
> but the counter tests for >= 0, which by default it starts as 0?

Reply:
Right make perfect sense, I thought it was strictly greater for some reason